docs/rfcs/0004-agent-flow.md

```diff
@@ -0,0 +1,54 @@
+# Agent Flow - Agent Utilizing Components 
+
+* Date: 2022-03-30
+* Author: Matt Durham (@mattdurham)
+* PR: [grafana/agent#1538](https://github.com/grafana/agent/pull/1538)
+* Status: Draft
+
+## Overarching Problem Statement
+
+The Agents configuration and onboarding is difficult to use. Viewing the effect of configuration changes on telemetry data is difficult. Making the configuration simplier, composable and intuitive allieviates these concerns.
```

Comment:
I am 90% aligned with the above. 

It feels like we are approaching the same solution from two different angles and the configuration is the core pivot point. My viewpoint is heavily user-driven, in that how do users gain more agency and understanding into the Agent. The primary tool that users have is via configuration so that is a focus. I think composability is the way to simplify and break down the problem into discreet steps to offer the user. The simplification of components leads to easier development of features and a natural pipeline approach is a great benefit.

Whereas, IMO (and please correct if wrong) you want composable components to make development easier and the system more flexible that also has a large benefit to the user. 

My goal is to build an argument that creating an inherent data flow diagram is a huge gain for the user, and allows a great user experience. And that the best way to do that is via composable components. I am unsure that our current configuration can do that. That even if we heavily changed the current configuration and made the underlying systems "composable" we would still fail from a user usability standpoint. IE extensions and pipelines the Agent uses for traces are good examples of systems that arent intuitive enough for my ideal end goal while still being pipelines. 

My thought is we want the same goal but the ordering of priorities might be different.

Reply:
I agree that it seems we want the same solution :) Composable components solve a lot of problems. They introduce some justified complexity but tooling can alleviate much of that pain. 

> Whereas, IMO (and please correct if wrong) you want composable components to make development easier and the system more flexible that also has a large benefit to the user.

I'm not so concerned about the development aspect (though composable components mean we can finally stop writing `thingManager`s. I'm definitely more focused on the power we're providing users: these components feel very unix-y to me, and I'm interested in how we can create a set of building blocks to unlock powerful and flexible flow graphs. 

> I am 90% aligned with the above.

What parts are you unaligned with? Should we discuss those?

Reply:
I think we are aligned enough to close this. There arent any fundamental differences.